corehq/apps/users/models_sql.py

```diff
@@ -16,12 +22,14 @@ class SQLUserRole(SyncSQLToCouchMixin, models.Model):
     # role can be assigned by all non-admins
     is_non_admin_editable = models.BooleanField(null=False, default=False)
     is_archived = models.BooleanField(null=False, default=False)
-    upstream_id = models.CharField(max_length=32, null=True)
+    upstream_id = models.IntegerField(null=True)
```

Comment:
Is this field still used in the context of remote linked domains? From [our discussion in the design doc](https://docs.google.com/document/d/1wVxmo-12dPIMwei2rn-6X9SL3GU8_KJLnY0VeVWlXdA/edit?disco=AAAAIbWWfys), it sounded like those roles would not exist in the local db, but having this be an integer referencing an auto-incrementing key seems to make the possibility of collisions between local ids and remote ids more likely. Is that a concern?

I didn't see any tests that mentioned remote linked domains. Could use of `upstream_id` in that scenario be tested?

Reply:
This is covered by tests here: https://github.com/dimagi/commcare-hq/blob/185d0ce3a99c95172014c20acb48ceeaeca2f7d3/corehq/apps/linked_domain/tests/test_update_roles.py#L9

Clashes with a local ID isn't an issue and will just result in a mapping from remote ID local roles that looks like this:
```
local_roles_by_upstream_id = {1: SQLRole(id=1)}
```

Reply:
I don't understand what you meant by "mapping from remote ID local roles." What is a "remote ID local role"?

I also don't see how those tests are verifying the scenario where a remote domain's roles (i.e., roles that are not in the local db) are mapped to a local domain's roles using `upstream_id`. My understanding is that both `self.domain` and `self.linked_domain` in those tests are always domains in the local db, and all of their roles exist in the local db. I expected the upstream (remote) domain to not exist in the local db.

Reply:
@millerdev this does not test that case because of an assumption I had made which is that there are no active remote linked domains that sync roles. I have since discovered that this is not the case, there are staging domains that have synced roles from prod. To avoid having to do a complex re-linking I am going to revert this to using the couch ID since that does not require a lookup to translate it to the SQL ID and so it can be treated as an opaque key.

Reply:
> "mapping from remote ID local roles."
This was meant to say "mapping from remote ID to local roles."

Reply:
Here's a test: 6ce726897e3 along with a change to how assignment mapping is done to prevent saving remote IDs before mapping them to local IDs.

Reply:
> because of an assumption I had made which is that there are no active remote linked domains that sync roles

In hindsight this was probably a very naive assumption :(